### PR TITLE
add waitForInitialization check

### PR DESCRIPTION
**Pull request by mmrj**
Continuation of #6 ... 

[SC-154168](https://app.shortcut.com/launchdarkly/story/154168/update-javascript-sdk-hello-app-readme-and-comments-to-standard-design), the Docs team is working on updating the Hello apps to [match the spec](https://launchdarkly.atlassian.net/wiki/spaces/ENG/pages/1499562073/SDK+hello+world).

Here, we're adding in the initialization check.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -34,14 +34,21 @@
 
       const ldclient = LDClient.initialize(clientSideID, user);
 
-      function render() {
-        const flagValue = ldclient.variation(flagKey, false);
-        const label = 'Feature flag ' + flagKey + ' is ' + flagValue + ' for this user';
-        div.replaceChild(document.createTextNode(label), div.firstChild);
-      }
+      ldclient.waitForInitialization().then(() => {
```
Comment (mmrj) [inline index.html]: I'm not sure if this is the correct way to do this... I'm attempting to follow instructions [in the reference docs](https://docs.launchdarkly.com/sdk/client-side/javascript#initializing-the-client) and [SDK docs](https://launchdarkly.github.io/js-client-sdk/interfaces/_launchdarkly_js_client_sdk_.ldclient.html#waitforinitialization).

File: index.html [hunk 2]
```diff
@@ -34,14 +34,21 @@
...
+        div.replaceChild(document.createTextNode('SDK successfully initialized!'), div.firstChild);
+
+        function render() {
+          const flagValue = ldclient.variation(flagKey, false);
+          const label = 'Feature flag ' + flagKey + ' is ' + flagValue + ' for this user';
+          div.replaceChild(document.createTextNode(label), div.firstChild);
+        }
+
+        render();
```
Comment (mmrj) [inline index.html]: Previously, this was `ldclient.on('ready', render);`. However, once we're in the `waitForInitialization` block, I had to call `render()` explicitly the first time to get the "Feature flag xx is true/false for this user" message to appear. So I'm not sure that this is correct.

**Review by scribblingfox — APPROVED** [review]
I'm approving this to get out of your way, but definitely hold off on merging until we can get a technical reviewer.

**Comment by kinyoklion** [comment]
I am a little confused on the goal. This hello example was waiting for the client to be ready, by using the `ready` callback, and then rendering. The promise versions are provided as variations which are more convenient for specific applications.

For instance, if you were using async code, you could do
```
  await client.waitUntilReady()
  render();
  client.on('change', render);
```

For this example though, because it is just basic JS and doesn't use async code currently, the ready callback a good use.

If we did want to use the promise, then you could do
```
client.WaitUntilReady().then(render);
```

I think this repo had some different problems in the linked document.

It doesn't log when there isn't an environment key set, some of the code needs name and should use const/etc.

**Comment by mmrj** [comment]
Hi @kinyoklion , thanks for your feedback and sorry to be unclear on this. 

> I think this repo had some different problems in the linked document.
> 
> It doesn't log when there isn't an environment key set, some of the code needs name and should use const/etc.

These were mostly fixed in #6, for example there’s now a [message when the clientSideID isn’t set](https://github.com/launchdarkly/hello-js/blob/9dae753fb9fd40e3a6be0feb3e4524b45365ce95/index.html#L31), and `const`s for the [clientSideID](https://github.com/launchdarkly/hello-js/blob/9dae753fb9fd40e3a6be0feb3e4524b45365ce95/index.html#L14), [flagKey](https://github.com/launchdarkly/hello-js/blob/9dae753fb9fd40e3a6be0feb3e4524b45365ce95/index.html#L17), and [user](https://github.com/launchdarkly/hello-js/blob/9dae753fb9fd40e3a6be0feb3e4524b45365ce95/index.html#L21). 

> This hello example was waiting for the client to be ready, by using the `ready` callback, and then rendering. 

Agreed. The one thing that it wasn’t doing but that the [spec requests (step 3)](https://launchdarkly.atlassian.net/wiki/spaces/ENG/pages/1499562073/SDK+hello+world#Standard-behavior) is to wait for initialization to complete and then show a message if so or if not (“SDK successfully initialized!” or “SDK failed to initialize”). 

So what I’m attempting to do in this PR is have that message appear. I’m not sure what the best way to do that is. Is there a way to display both success and failure messages using either the `ready` callback or the `client.WaitUntilReady().then(render);` ?

**Comment by eli-darkly** [comment]
> Is there a way to display both success and failure messages using either the ready callback or the `client.WaitUntilReady().then(render);`?

If I understand correctly, the issue here is just the failure message. The success message is already addressed by the current code that uses the `ready` event. The equivalent, _if_ we want to retain the current event-based implementation, would be `client.on('failed', [function name or anonymous function to call if initialization failed])`. In other words, the `ready` event is equivalent to the `waitForInitialization` promise resolving, and the `failed` event is equivalent to the `waitForInitialization` promise rejecting. There's a list of all these events [here](https://launchdarkly.github.io/js-client-sdk/interfaces/_launchdarkly_js_client_sdk_.ldclient.html#on).

Whether it's better/clearer for this example code to use events or to use promises, I'm not sure. But we can accomplish the same things either way.

**Comment by kinyoklion** [comment]
@mmrj Sorry for my lack of reply. I github'd poorly. I think Eli covers what would need to be done.

I am neutral to callbacks versus promises here, but I think the existing code isn't bad, because it is pretty easy to digest what is going on. I do think if we use promises, we should probably use them from an `async` method and use `await`.

**Comment by mmrj** [comment]
Thank you both! I'll update to add the `failed` event and keep this more similar to the original code.

**Comment by mmrj** [comment]
@eli-darkly Okay, I think I've updated correctly based on your suggestions. Ready for a re-review at your convenience.

**Review by eli-darkly — APPROVED** [review]

**Pull request merged**